File: in_Java/0012 Integer to Roman.java
class Solution {
    public String intToRoman(int num) {
        Map<Integer,String> hm = new HashMap<>();
        String[] arr1 = new String[]{"I","V","X","L","C","D","M","IV","IX","XL","XC","CD","CM"};
        int[] arr2 = new int[]{1,5,10,50,100,500,1000,4,9,40,90,400,900};
        for(int i = 0; i < arr1.length; i++){
            hm.put(arr2[i],arr1[i]);
        }
        Map<Integer, int[]> hm2 = new HashMap<>();
        hm2.put(1, new int[]{1});
        hm2.put(2, new int[]{1,1});
        hm2.put(3, new int[]{1,1,1});
        hm2.put(4, new int[]{4});
        hm2.put(5, new int[]{5});
        hm2.put(6, new int[]{5,1});
        hm2.put(7, new int[]{5,1,1});
        hm2.put(8, new int[]{5,1,1,1});
        hm2.put(9, new int[]{9});
        
        String res = "";
        String cur;
        int t;
        int id = 0;
        int idx;
        while(num != 0){
            t = num % 10;
            if(t == 0){
                num /= 10;
                id++;
                continue;
            }
            int[] x = hm2.get(t);
            cur = "";
            for(int i : x){
                //System.out.println(i);
                idx = (int)(i * Math.pow(10,id));
                cur += hm.get(idx);
            }
            res = cur + res;
            num /= 10;
            id++;
        }
        return res;
    }
}

File: in_Python/0012 Integer to Roman.py
class Solution(object):
    def intToRoman(self, num):
        """
        :type num: int
        :rtype: str
        """
        d = {}
        d[1] = "I"
        d[5] = "V"
        d[10] = "X"
        d[50] = "L"
        d[100] = "C"
        d[500] = "D"
        d[1000] = "M"
        d[4] = "IV"
        d[9] = "IX"
        d[40] = "XL"
        d[90] = "XC"
        d[400] = "CD"
        d[900] = "CM"
        res = []
        ks = [1000,900,500,400,100,90,50,40,10,9,5,4,1]
        while num != 0:
            for k in ks:
                if k <= num:
                    res.append(d[k]) 
                    num -= k
                    break
        return ''.join(res)
                